pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,526 @@
+			return fmt.Errorf("Unexpected error getting new provisioner for QoS Class %s: %v\n", qos, err)
+		}
+		newVolume, err := provisioner.NewPersistentVolumeTemplate()
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting new volume template for QoS Class %s: %v\n", qos, err)
+		}
+
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference for %s: %v\n", claim.Name, err)
+		}
+
+		// the creation of this volume is the bind to the claim.
+		// The claim will match the volume during the next sync period when the volume is in the local cache
+		newVolume.Spec.ClaimRef = claimRef
+		newVolume.Annotations[provisionedForKey] = ClaimToProvisionableKey(claim)
+		newVolume.Annotations[pvProvisioningRequired] = "true"
+		newVolume.Annotations[qosProvisioningKey] = qos
+		newVolume, err = controller.client.CreatePersistentVolume(newVolume)
+		glog.V(5).Infof("PersistentVolume[%s] created for PVC[%s]", newVolume.Name, claim.Name)
+		if err != nil {
+			return fmt.Errorf("PersistentVolumeClaim[%s] failed provisioning: %+v", claim.Name, err)
+		}
+
+		claim.Annotations[pvProvisioningRequired] = pvProvisioningCompleted
+		_, err = controller.client.UpdatePersistentVolumeClaim(claim)
+		if err != nil {
+			glog.Error("error updating persistent volume claim: %v", err)
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("No provisioner found for qos request: %s", qos)
+}
+
+func (controller *PersistentVolumeController) reconcileVolume(pv *api.PersistentVolume) error {
+	glog.V(5).Infof("PersistentVolume[%s] reconciling", pv.Name)
+
+	if pv.Spec.ClaimRef == nil {
+		glog.V(5).Infof("PersistentVolume[%s] is not bound to a claim.  No provisioning required", pv.Name)
+		return nil
+	}
+
+	// TODO:  fix this leaky abstraction.  Had to make our own store key because ClaimRef fails the default keyfunc (no Meta on object).
+	obj, exists, _ := controller.claimStore.GetByKey(fmt.Sprintf("%s/%s", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name))
+	if !exists {
+		return fmt.Errorf("PersistentVolumeClaim[%s/%s] not found in local cache", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)
+	}
+
+	claim := obj.(*api.PersistentVolumeClaim)
+
+	// no provisioning required, volume is ready and Bound
+	if !keyExists(pvProvisioningRequired, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] does not require provisioning", pv.Name)
+		return nil
+	}
+
+	// provisioning is completed, volume is ready and return Bound
+	if isAnnotationMatch(pvProvisioningRequired, pvProvisioningCompleted, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] is bound and provisioning is complete", pv.Name)
+		provisionedFor, _ := pv.Annotations[provisionedForKey]
+		if provisionedFor != ClaimToProvisionableKey(claim) {
+			return fmt.Errorf("pre-bind mismatch - expected %s but found %s", ClaimToProvisionableKey(claim), provisionedFor)
+		}
+		return nil
+	}
+
+	// provisioning is not complete, return Pending
```

Comment:
The comment is quite confusing, as nothing 'Pending' is returned. Is 'return <new state>' missing?